internal/util/errors.go

```diff
@@ -16,98 +16,46 @@ limitations under the License.
-}
-
-// ErrObjectExists is returned when named omap is already present in rados.
-type ErrObjectExists struct {
-	objectName string
-	err        error
-}
-
-// Error returns the error string for ErrObjectExists.
-func (e ErrObjectExists) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrObjectExists) Unwrap() error {
-	return e.err
-}
-
-// ErrObjectNotFound is returned when named omap is not found in rados.
-type ErrObjectNotFound struct {
-	oMapName string
-	err      error
-}
-
-// Error returns the error string for ErrObjectNotFound.
-func (e ErrObjectNotFound) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrObjectNotFound) Unwrap() error {
-	return e.err
-}
-
-// ErrSnapNameConflict is generated when a requested CSI snap name already exists on RBD but with
-// different properties, and hence is in conflict with the passed in CSI volume name.
-type ErrSnapNameConflict struct {
-	requestName string
-	err         error
-}
+import (
+	"errors"
+	"fmt"
+)
 
-// Error returns the error string for ErrSnapNameConflict.
-func (e ErrSnapNameConflict) Error() string {
-	return e.err.Error()
-}
-
-// Unwrap returns the encapsulated error.
-func (e ErrSnapNameConflict) Unwrap() error {
-	return e.err
-}
+var (
+	// ErrKeyNotFound is returned when requested key in omap is not found.
+	ErrKeyNotFound = errors.New("key not found")
+	// ErrObjectExists is returned when named omap is already present in rados.
+	ErrObjectExists = errors.New("object exists")
+	// ErrObjectNotFound is returned when named omap is not found in rados.
+	ErrObjectNotFound = errors.New("object not found")
+	// ErrSnapNameConflict is generated when a requested CSI snap name already exists on RBD but with
+	// different properties, and hence is in conflict with the passed in CSI volume name.
+	ErrSnapNameConflict = errors.New("snapshot name conflict")
+	// ErrPoolNotFound is returned when pool is not found.
+	ErrPoolNotFound = errors.New("pool not found")
+)
 
-// NewErrSnapNameConflict returns a ErrSnapNameConflict error when CSI snap name already exists.
-func NewErrSnapNameConflict(name string, err error) ErrSnapNameConflict {
-	return ErrSnapNameConflict{name, err}
+type errorPair struct {
+	first  error
+	second error
 }
 
-// ErrPoolNotFound is returned when pool is not found.
-type ErrPoolNotFound struct {
-	Pool string
-	Err  error
+func (e errorPair) Error() string {
+	return fmt.Sprintf("%v: %v", e.first, e.second)
 }
 
-// Error returns the error string for ErrPoolNotFound.
-func (e ErrPoolNotFound) Error() string {
-	return e.Err.Error()
+// Is checks if target error is wrapped in the first error.
+func (e errorPair) Is(target error) bool {
+	return errors.Is(e.first, target)
 }
 
-// Unwrap returns the encapsulated error.
-func (e ErrPoolNotFound) Unwrap() error {
-	return e.Err
+// Unwrap returns the second error.
+func (e errorPair) Unwrap() error {
+	return e.second
```

Comment:
I see `Unwrap()` as a function to convert one error into an other. That makes it confusing to get the 2nd error back with `Unwrap()` (my thinking is something along primary, secondary error).

What is your motivation to use `first` and `second`? If I can follow your line of thought, it might make more sense to me.

Reply:
An `errorPair` in fact wraps two errors, with no further meaning. So the `Unwrap()` method must return one of them. It doesn't matter which one. And the `Is()` must check for the other one. So that both branches are searched for the given Error.

As I said before, I just took the naming from C++ pairs, where they are also called first and second, without any further meaning. The only "meaning" in this case would be, that when converted to a string, it concatenates them like `first.Error() + ": " + second.Error()`. Also, when doing an `errors.Is()`, the `first` error would be checked first (by using `Is()`), after that it would check `second` (by using `Unwrap()`). That's just how `errors.Is()` is implemented, that it first uses `Is()` and then `Unwrap()`. That's all. In my first patch-set I called it `left` and `right`. Would that work better for you?

Reply:
Ok, thanks for the explanation. I did not check how `errors.Is()` and `errors.As()` are implemented, so with these details it should be good to go :+1: